### ISSUE TITLE
Rule proposal: Enforce read-only

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#514]
I would like to enforce function arguments to be read-only (immutable), to prevent me from accidentally mutate user input. I've encountered many modules that accidentally modify a user-supplied options-object.

I would also like to enforce function return values to *not* be read-only, as that's annoying for users without much benefit, since it's not actually enforced at runtime. This one might be slightly controversial, so could be an option.

## Fail

```ts
function foo(input: number[]): void {}

function foo(input: string): Readonly<Point> {}
```

## Pass

```ts
function foo(input: readonly number[]): void {}

function foo(input: string): Point {}
```

The rule should support enforcing read-only for:
- [`readonly` (array and tuple)](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#improvements-for-readonlyarray-and-readonly-tuples)
- [`readonly` (object property)](https://www.typescriptlang.org/docs/handbook/interfaces.html#readonly-properties)
- [`Readonly` (object properties)](https://basarat.gitbooks.io/typescript/docs/types/readonly.html)
- `ReadonlyArray`
- `readonly` for object keys: `{readonly [user: string]: boolean}`

**Comment by kitsonk** [comment]
I would like to see the ability to allow the return of read-only values in function arguments (so optional).  While it can be annoying, if the internals of a function have retained references to objects/arrays that you expect to be immutable, then you would want to be able to express that intent of "look, but don't touch, or it might get ugly".

**Comment by bradzacher** [comment]
I was having a look into this (because it's one of our highest-voted issues), and it should be simple enough to do.
It will require type information though, so that the rule can properly resolve the types to determine if they're readonly:

```ts
// file1.ts
export type MyReadonlyType = readonly string[];

// file2.ts
import { MyReadonlyType } from './file1';

function foo(arg: MyReadonlyType) {}
```

**Comment by bradzacher** [comment]
Hmmm. Should this check be deeply nested, or just top level?

I.e. would the following types be considered readonly for the purpose of the rule?

```
interface DeepObject {
  readonly one: {
    two: string;
  };
}

type DeepArray = readonly (string[])[];
```

**Comment by sindresorhus** [comment]
> Should this check be deeply nested, or just top level?

I would prefer it to check deeply nested.

**Comment by bradzacher** [comment]
Should methods be treated as "readonly"?
There's no way to create a readonly method, only readonly class properties with function values.

```ts
interface Foo {
  bar(): string;
}
declare const x: Foo;

// this is legal, but people "don't" do it
x.bar = () => 'aaa';
```
